wasm.md

```diff
@@ -638,6 +652,101 @@ Unlike labels, only one frame can be "broken" through at a time.
+
+```k
+    syntax Instr ::= "(" "memory"               ")"
+                   | "(" "memory" Int           ")" // Max only
+                   | "(" "memory" Int Int       ")" // Min and max.
+                   |     "memory" Int OptionInt
+ // --------------------------------------------
+    rule <k> ( memory         ) => memory 0   none     ... </k>
+    rule <k> ( memory     MAX ) => memory 0   some MAX ... </k>
+    rule <k> ( memory MIN MAX ) => memory MIN some MAX ... </k>
+
+    rule <k> memory _:Int _:OptionInt => trap ... </k>
+         <memAddrs> some _ </memAddrs>
+
+    rule <k> memory MIN MAX:OptionInt => . ... </k>
+         <memAddrs>    none => some NEXT </memAddrs>
+         <nextMemAddr> NEXT => NEXT +Int 1 </nextMemAddr>
+         <mems>
+           ( .Bag
+          => <memInst>
+               <memAddr> NEXT </memAddr>
+               <mmax>    MAX  </mmax>
+               <msize>   MIN  </msize>
+             </memInst>
+           )
+           ...
+         </mems>
+```
+
+TODO: There are many more valid ways to instantiate memory. Allow specifying just max or neither min or max, folded syntax, identifier, inline instantiation, and inline export and import.
+
+
+Memory instances can be either bounded by a max number of pages, or unbounded.
+However, the absolute max allowed size if 2^16 pages.
+Incidentally, the page size is 2^16 bytes.
+
+```k
+    syntax Int ::= #pageSize()      [function]
+    syntax Int ::= #maxMemorySize() [function]
+ // ------------------------------------------
+    rule #maxMemorySize() => 65536
+    rule #pageSize()      => 65536
+```
+
+The `size` operation return the size of the memory, measured in pages.
+
+```k
+    syntax Instr ::= "(" "memory.size" ")"
+ // --------------------------------------
+    rule <k> ( memory.size ) => < i32 > SIZE ... </k>
+         <memAddrs> some ADDR </memAddrs>
+         <memInst>
+           <memAddr> ADDR </memAddr>
+           <msize>   SIZE </msize>
+           ...
+         </memInst>
+```
+
+`grow` increases the size of memory in units of pages.
+Failure to grow is indicated by pushing -1 to the stack.
+Success is indicated by pushing the previous memory size to the stack.
+`grow` is non-deterministic and may fail either due to trying to exceed explicit max values, or because the embedder does not have resources available.
+By setting the `<deterministicMemoryGrowth>` field in the configuration to `true`, the sematnics ensure memory growth only fails if the memory in question would exceed max bounds.
+
+```k
+    syntax Bool  ::= #growthAllowed(Int, OptionInt) [function]
+    syntax Instr ::= "(" "memory.grow" ")" | "(" "memory.grow" Instr ")" | "grow" Int
+ // ---------------------------------------------------------------------------------
+    rule #growthAllowed(SIZE, none  ) => SIZE <=Int #maxMemorySize()
+    rule #growthAllowed(SIZE, some I) => #growthAllowed(SIZE, none) andBool SIZE <=Int I
```

Comment:
I usually put the semantics of helper functions below top-level semantic rules, so that people can gloss over them if they can guess what they do from the name anyway:

Eg, I would structure like:

```
syntax Instr ::= ...
-----------------
RULES for memory

syntax Bool ::= myHelperPred
----------------------
RULE FOR HELPER PRED
```

Reply:
Since sizes and maxes are common to all the existing memory instructions, I put them last in the memory section, which I think works well.